### ISSUE TITLE
Cannot set com.apple.loginwindow.SHOWOTHERUSERS_MANAGED through system.defaults.CustomSystemPreferences

### ISSUE THREAD

**Issue by ciferkey** [LnL7/nix-darwin#614]
I'm trying to use [system.defaults.CustomSystemPreferences](https://daiderd.com/nix-darwin/manual/index.html#opt-system.defaults.CustomSystemPreferences) to [disable](https://apple.stackexchange.com/a/303271) the Other Users login option.

This should involve setting `com.apple.loginwindow.SHOWOTHERUSERS_MANAGED` to false:

```nix
﻿﻿  system.defaults.CustomSystemPreferences = {
    com.apple.loginwindow = {
      com.apple.loginwindow.SHOWOTHERUSERS_MANAGED = true;
    };
  };
```

However when I try to `darwin-rebuild` I get the error:
```
error: invalid value type
```

<details>
  <summary>Full --show-trace</summary>

  ```
error:
       … while evaluating the attribute 'activationScript' of the derivation 'darwin-system-23.05.20230309.1e25906+darwin4.87b9d09'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/pkgs/stdenv/generic/make-derivation.nix:294:7:

          293|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          294|       name =
             |       ^
          295|         let

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling 'g'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/attrsets.nix:535:19:

          534|           g =
          535|             name: value:
             |                   ^
          536|             if isAttrs value && cond value

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/attrsets.nix:538:20:

          537|               then recurse (path ++ [name]) value
          538|               else f (path ++ [name]) value;
             |                    ^
          539|         in mapAttrs g;

       … while calling anonymous lambda

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:274:72:

          273|           # For definitions that have an associated option
          274|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          275|

       … while evaluating the attribute 'value'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:752:9:

          751|     in warnDeprecation opt //
          752|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          753|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.activationScripts.script.text':

       … while evaluating the attribute 'mergedValue'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:787:5:

          786|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          787|     mergedValue =
             |     ^
          788|       if isDefined then

       … while evaluating the attribute 'values'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:781:9:

          780|       in {
          781|         values = defs''';
             |         ^
          782|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:880:7:

          879|     in {
          880|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          881|       inherit highestPrio;

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:767:17:

          766|         # Process mkMerge and mkIf properties.
          767|         defs' = concatMap (m:
             |                 ^
          768|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while calling anonymous lambda

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:767:28:

          766|         # Process mkMerge and mkIf properties.
          767|         defs' = concatMap (m:
             |                            ^
          768|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating definitions from `/nix/store/yw0f0bpl89ryj7cv1c7dwbw0g0j0mcfi-source/modules/system/activation-scripts.nix':

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:768:137:

          767|         defs' = concatMap (m:
          768|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          769|         ) defs;

       … while calling 'dischargeProperties'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:839:25:

          838|   */
          839|   dischargeProperties = def:
             |                         ^
          840|     if def._type or "" == "merge" then

       … while evaluating the attribute 'value'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:592:44:

          591|       defnsByName' = byName "config" (module: value:
          592|           [{ inherit (module) file; inherit value; }]
             |                                            ^
          593|         ) configs;

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling 'g'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/attrsets.nix:535:19:

          534|           g =
          535|             name: value:
             |                   ^
          536|             if isAttrs value && cond value

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/attrsets.nix:538:20:

          537|               then recurse (path ++ [name]) value
          538|               else f (path ++ [name]) value;
             |                    ^
          539|         in mapAttrs g;

       … while calling anonymous lambda

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:274:72:

          273|           # For definitions that have an associated option
          274|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          275|

       … while evaluating the attribute 'value'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:752:9:

          751|     in warnDeprecation opt //
          752|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          753|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.activationScripts.defaults.text':

       … while evaluating the attribute 'mergedValue'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:787:5:

          786|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          787|     mergedValue =
             |     ^
          788|       if isDefined then

       … while evaluating the attribute 'values'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:781:9:

          780|       in {
          781|         values = defs''';
             |         ^
          782|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:880:7:

          879|     in {
          880|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          881|       inherit highestPrio;

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:767:17:

          766|         # Process mkMerge and mkIf properties.
          767|         defs' = concatMap (m:
             |                 ^
          768|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while calling anonymous lambda

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:767:28:

          766|         # Process mkMerge and mkIf properties.
          767|         defs' = concatMap (m:
             |                            ^
          768|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating definitions from `/nix/store/yw0f0bpl89ryj7cv1c7dwbw0g0j0mcfi-source/modules/system/defaults-write.nix':

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:768:137:

          767|         defs' = concatMap (m:
          768|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          769|         ) defs;

       … while calling 'dischargeProperties'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:839:25:

          838|   */
          839|   dischargeProperties = def:
             |                         ^
          840|     if def._type or "" == "merge" then

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:845:11:

          844|         if def.condition then
          845|           dischargeProperties def.content
             |           ^
          846|         else

       … while calling 'dischargeProperties'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:839:25:

          838|   */
          839|   dischargeProperties = def:
             |                         ^
          840|     if def._type or "" == "merge" then

       … while evaluating the attribute 'content'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/modules.nix:910:14:

          909|     { _type = "if";
          910|       inherit condition content;
             |              ^
          911|     };

       … while evaluating call site

       at /nix/store/yw0f0bpl89ryj7cv1c7dwbw0g0j0mcfi-source/modules/system/defaults-write.nix:43:29:

           42|   CustomUserPreferences = flatten (mapAttrsToList (name: value: defaultsToList name value) cfg.CustomUserPreferences);
           43|   CustomSystemPreferences = flatten (mapAttrsToList (name: value: defaultsToList name value) cfg.CustomSystemPreferences);
             |                             ^
           44|

       … while calling 'flatten'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:138:13:

          137|   */
          138|   flatten = x:
             |             ^
          139|     if isList x

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:140:10:

          139|     if isList x
          140|     then concatMap (y: flatten y) x
             |          ^
          141|     else [x];

       … while calling anonymous lambda

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:140:21:

          139|     if isList x
          140|     then concatMap (y: flatten y) x
             |                     ^
          141|     else [x];

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:140:24:

          139|     if isList x
          140|     then concatMap (y: flatten y) x
             |                        ^
          141|     else [x];

       … while calling 'flatten'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:138:13:

          137|   */
          138|   flatten = x:
             |             ^
          139|     if isList x

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:140:10:

          139|     if isList x
          140|     then concatMap (y: flatten y) x
             |          ^
          141|     else [x];

       … while calling anonymous lambda

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:140:21:

          139|     if isList x
          140|     then concatMap (y: flatten y) x
             |                     ^
          141|     else [x];

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:140:24:

          139|     if isList x
          140|     then concatMap (y: flatten y) x
             |                        ^
          141|     else [x];

       … while calling 'flatten'

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/lists.nix:138:13:

          137|   */
          138|   flatten = x:
             |             ^
          139|     if isList x

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling anonymous lambda

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/attrsets.nix:479:10:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |          ^
          480|

       … while evaluating call site

       at /nix/store/mc4rhg6n6zy7i3wifrxva08lgz4q29nn-source/lib/attrsets.nix:479:16:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |                ^
          480|

       … while calling 'writeDefault'

       at /nix/store/yw0f0bpl89ryj7cv1c7dwbw0g0j0mcfi-source/modules/system/defaults-write.nix:17:31:

           16|
           17|   writeDefault = domain: key: value:
             |                               ^
           18|     "defaults write ${domain} '${key}' ${writeValue value}";

       … while evaluating call site

       at /nix/store/yw0f0bpl89ryj7cv1c7dwbw0g0j0mcfi-source/modules/system/defaults-write.nix:18:42:

           17|   writeDefault = domain: key: value:
           18|     "defaults write ${domain} '${key}' ${writeValue value}";
             |                                          ^
           19|

       … while calling 'writeValue'

       at /nix/store/yw0f0bpl89ryj7cv1c7dwbw0g0j0mcfi-source/modules/system/defaults-write.nix:10:16:

            9|
           10|   writeValue = value:
             |                ^
           11|     if isBool value then "-bool ${boolValue value}" else

       error: invalid value type
  ```
</details>

**Comment by 2gn** [comment]
From what I see in the [example](https://github.com/LnL7/nix-darwin/blob/3db1d870b04b13411f56ab1a50cd32b001f56433/modules/system/defaults/CustomPreferences.nix), I think you need to do like this:

```darwin-configuration.nix
{
  system.defaults.CustomSystemPreferences = {
    "com.apple.loginwindow" = {
      "com.apple.loginwindow.SHOWOTHERUSERS_MANAGED" = true;
    };
  };
}
```

**Comment by ciferkey** [comment]
Thank you. That built without error!

I noticed in the example I had it send to `true`. So also updated it to `false`. It doesn't seem to actually disable the other user log in option thought but that looks like a secondary issue. Running the command from the stackexchange post does disable it however.

**Comment by 2gn** [comment]
```zsh
Activating linkGeneration
Cleaning up orphan links from /Users/user
No change so reusing latest profile generation 75
Creating home file links in /Users/user
Activating onFilesChange
Activating setupLaunchAgents
user@computer> sudo defaults read /Library/Preferences/com.apple.loginwindow
{
    AccountInfo =     {
        FirstLogins =         {
            user = 1;
        };
        MaximumUsers = 1;
        OnConsole =         {
        };
    };
    GuestEnabled = 0;
    OptimizerLastRunForBuild = 46336128;
    OptimizerLastRunForSystem = 218235136;
    UseVoiceOverLegacyMigrated = 1;
    lastUser = loggedIn;
    lastUserName = user;
}
```

You're right. Defaults value was not modified. Maybe there are some bugs around [here](https://github.com/LnL7/nix-darwin/blob/3db1d870b04b13411f56ab1a50cd32b001f56433/modules/system/defaults-write.nix#L43), or I may have written my configuration wrong.

**Comment by han-tyumi** [comment]
I believe support would need to be added for dictionary values [here](https://github.com/LnL7/nix-darwin/blob/0dd382b70c351f528561f71a0a7df82c9d2be9a4/modules/system/defaults-write.nix#L10).
I recently ran into this myself with a different settings.

**Comment by Enzime** [comment]
Adding support for dictionaries should be fixed by #897
